src/test/java/htsjdk/samtools/util/StringUtilTest.java

```diff
@@ -67,4 +67,53 @@ public void testSplit(final String input, final String[] expectedResult, final b
                 {"A:BB:C:", new String[]{"A", "BB", "C:"}, true}, 
         };
     }
+
+    @DataProvider(name="withinHammingDistanceProvider")
+    public Object[][] isWithinHammingDistanceProvider() {
+        return new Object[][] {
+                {"ATAC", "GCAT", 3, true},
+                {"ATAC", "GCAT", 2, false},
+                {"ATAC", "GCAT", 1, false},
+                {"ATAC", "GCAT", 0, false}
+        };
+    }
+
+    @Test(dataProvider = "withinHammingDistanceProvider")
+    public void testIsWithinHammingDistance(final String s1, final String s2, final int maxHammingDistance, final boolean expectedResult) {
+        Assert.assertEquals(StringUtil.isWithinHammingDistance(s1, s2, maxHammingDistance), expectedResult);
+    }
+
+    @DataProvider(name="withinHammingDistanceExceptionProvider")
+    public Object[][] isWithinHammingDistanceException() {
+        return new Object[][] {
+                {"ATAC", "GCT", 3, true},
+                {"ATAC", "AT", 2, false},
+                {"ATAC", "T", 1, false},
+                {"", "GCAT", 0, false}
+        };
+    }
+
+    @Test(dataProvider = "hammingDistanceExceptionProvider", expectedExceptions = IllegalArgumentException.class)
+    public void testIsWithinHammingDistanceExceptions(final String s1, final String s2, final int maxHammingDistance) {
+        // We assert hammingDistance = 0, and isWithinHammingDistance = true because the values don't matter
+        // and we are checking to ensure that the IllegalArgumentException is thrown
+        Assert.assertEquals(StringUtil.hammingDistance(s1, s2), 0);
+        Assert.assertEquals(StringUtil.isWithinHammingDistance(s1, s2, maxHammingDistance), true);
+    }
+
+    @DataProvider(name="hammingDistanceProvider")
+    public Object[][] hammingDistance() {
+        return new Object[][] {
+                {"ATAC", "GCAT", 3},
+                {"ATAGC", "ATAGC", 0},
+                {"ATAC", "atac", 4}, // Hamming distance is case sensitive
```

Comment:
Lets defer adding support for matching valid bases for now, but just know that it may become useful in the future.